packages/backend/src/core/totalSupply/TotalSupplyUpdater.ts

```diff
@@ -0,0 +1,151 @@
+      'ChainId mismatch between totalSupplyProvider and totalSupplyUpdater provider',
+    )
+  }
+
+  async getTotalSuppliesWhenReady(
+    timestamp: UnixTime,
+    refreshIntervalMs = 1000,
+  ) {
+    while (!this.knownSet.has(timestamp.toNumber())) {
+      await setTimeout(refreshIntervalMs)
+    }
+    return this.totalSupplyRepository.getByTimestamp(this.chainId, timestamp)
+  }
+
+  async start() {
+    const known = await this.totalSupplyStatusRepository.getByConfigHash(
+      this.chainId,
+      this.configHash,
+    )
+
+    for (const timestamp of known) {
+      this.knownSet.add(timestamp.toNumber())
+    }
+
+    this.logger.info('Started')
+    return this.clock.onEveryHour((timestamp) => {
+      if (!this.knownSet.has(timestamp.toNumber())) {
+        // we add to front to sync from newest to oldest
+        this.taskQueue.addToFront(timestamp)
+      }
+    })
+  }
+
+  async update(timestamp: UnixTime) {
+    this.logger.debug('Update started', {
+      timestamp: timestamp.toNumber(),
+      chainId: this.chainId.toString(),
+    })
+    const known = await this.totalSupplyRepository.getByTimestamp(
+      this.chainId,
+      timestamp,
+    )
+
+    const missing = getMissingData(timestamp, known, this.tokens)
+
+    if (missing.length > 0) {
+      const blockNumber = await this.blockNumberUpdater.getBlockNumberWhenReady(
+        timestamp,
+      )
+
+      assert(blockNumber, 'No timestamp for this block number')
+
+      const totalSupplies = await this.totalSupplyProvider.getTotalSupplies(
+        missing,
+        timestamp,
+        blockNumber,
+      )
+
+      await this.totalSupplyRepository.addOrUpdateMany(totalSupplies)
+      this.logger.debug('Updated total supplies', {
+        timestamp: timestamp.toNumber(),
+        chainId: this.chainId.toString(),
+      })
+    } else {
+      this.logger.debug('Skipped updating total supplies', {
+        timestamp: timestamp.toNumber(),
+        chainId: this.chainId.toString(),
+      })
+    }
+    this.knownSet.add(timestamp.toNumber())
+    await this.totalSupplyStatusRepository.add({
+      chainId: this.chainId,
+      configHash: this.configHash,
+      timestamp,
+    })
+    this.logger.info('Update completed', { timestamp: timestamp.toNumber() })
+  }
+}
+
+export function getMissingData(
+  timestamp: UnixTime,
+  known: TotalSupplyRecord[],
+  tokens: TotalSupplyTokensConfig[],
+): TotalSupplyQuery[] {
+  const knownSet = new Set(
+    // FIXME: here? how to distinguish properly?
+    known.map(({ assetId }) => `${assetId.toString()}`),
+  )
+
+  const missing: TotalSupplyQuery[] = []
+
+  for (const token of tokens) {
+    if (token.sinceTimestamp.gt(timestamp)) {
+      continue
+    }
+
+    const entry: TotalSupplyQuery = {
+      tokenAddress: token.tokenAddress,
+      assetId: token.assetId,
+    }
+
+    if (!knownSet.has(`${entry.assetId.toString()}`)) {
+      missing.push(entry)
+    }
+  }
+
+  return missing
+}
```

Comment:
Check whether I ported logic properly since I may miss few things without going in-depth here

Reply:
tbh probably the tests will truly show, but looks okay

Reply:
https://github.com/l2beat/l2beat/pull/1664/commits/9f4aef37401e6b4df9e9c65e0fa231970e6ea3b6

Reply:
also slightly refactored https://github.com/l2beat/l2beat/pull/1664/commits/68a9f37a439e778098fd0c674ddc392ca5fb7f2b